doc/design/csp.md

```diff
@@ -0,0 +1,96 @@
+# Design Doc: CSP in PaddlePaddle Fluid
+
+## Motivations
+
+Concurrent programming is important for deep learning.  Example applications include
+
+1. A thread uses the GPU for computing while the main thread keeps loading the next minibatch, and
+1. a thread uploads the local gradients to the parameter server while the main thread keeps computing.
+
+Most DL systems, including TensorFlow, Caffe2, and MxNet, can asynchronously execute operators in a graph. However, Fluid doesn't have the concept graph at all, as the design goal of Fluid is a programming language.
```

Comment:
Just a note: the big majority of the TensorFlow OP is synchronous, the TF executor runs none-dependent OPs in parallel on different threads.

Reply:
Thanks for the note!